src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -12,202 +12,390 @@
     ) -> Option<CompactionInput> {
         let select_level = 0;
-        let target_level = self.target_level;
+        let target_level = self.target_level as u32;
         let next_task_id = self.compact_task_id;
 
-        if levels[select_level].table_infos.is_empty() {
+        let l0 = levels.l0.as_ref().unwrap();
+        if l0.sub_levels.is_empty()
+            || l0.sub_levels[0].level_type != LevelType::Nonoverlapping as i32
+        {
             return None;
         }
 
-        let (select_level_inputs, target_level_inputs) = self.select_input_files(
-            &levels[select_level],
-            &levels[target_level],
-            &level_handlers[select_level],
-            &level_handlers[target_level],
-        );
-        if select_level_inputs.is_empty() {
+        let is_l0_pending_compact = level_handlers[0].is_level_pending_compact(&l0.sub_levels[0]);
+
+        if !is_l0_pending_compact && levels.levels[self.target_level - 1].table_infos.is_empty() {
+            level_handlers[select_level].add_pending_task(
+                next_task_id,
+                self.target_level,
+                &l0.sub_levels[0].table_infos,
+            );
+            return Some(CompactionInput {
+                input_levels: vec![
+                    InputLevel {
+                        level_idx: 0,
+                        level_type: LevelType::Nonoverlapping as i32,
+                        table_infos: l0.sub_levels[0].table_infos.clone(),
+                    },
+                    InputLevel {
+                        level_idx: target_level as u32,
+                        level_type: LevelType::Nonoverlapping as i32,
+                        table_infos: vec![],
+                    },
+                ],
+                target_level: self.target_level,
+                target_sub_level_id: 0,
+            });
+        }
+
+        let mut input_levels = self.pick_min_overlap_tables(l0, &levels.levels, level_handlers);
+
+        if input_levels.is_empty() {
             return None;
         }
 
-        level_handlers[select_level].add_pending_task(next_task_id, &select_level_inputs);
-        level_handlers[target_level].add_pending_task(next_task_id, &target_level_inputs);
-        // Here, we have known that `select_level_input` is valid
-        let mut splits = Vec::with_capacity(target_level_inputs.len());
-        splits.push(KeyRange::new(vec![], vec![]));
-        if target_level_inputs.len() > 1 {
-            for table in &target_level_inputs[1..] {
-                let key_before_last = FullKey::from_user_key_slice(
-                    user_key(&table.key_range.as_ref().unwrap().left),
-                    HummockEpoch::MAX,
-                )
-                .into_inner();
-                splits.last_mut().unwrap().right = key_before_last.clone();
-                splits.push(KeyRange::new(key_before_last, vec![]));
+        const MAX_WRITE_AMPLIFICATION: u64 = 150;
```

Comment:
Should this be configurable via compaction config?

Reply:
🤔 It seems not necessary. Because in most case we will trigger a full compaction because the file does not split as table_id